### PR TITLE
Change .groupby fastpath to work for monotonic increasing and decreasing

### LINKED ISSUE #6220: [FEATURE]: Use fast path when grouping by unique monotonic decreasing variable
### Is your feature request related to a problem?

See https://github.com/pydata/xarray/pull/6213/files#r795716713

We check whether the `by` variable for groupby is unique and monotonically increasing. But the fast path would also apply to unique and monotonically decreasing variables.

### Describe the solution you'd like

Update the condition to `is_monotonic_increasing or is_monotonic_decreasing` and add a test.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by JoelJaeschke**
This fixes GH6220 which makes it possible to use the fastpath for .groupby for monotonically increasing and decreasing values.

- [x] Closes #6220
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

### CODE REVIEW

**Comment by dcherian** [comment]
Thanks @JoelJaeschke The intent was to change the code in `_unique_and_monotonic`.

A good test would be make sure you get the same answer when grouping a DataArray of ones by a monotonically increasing variable (e.g. `[1, 1, 2, 2, 3, 3, 3]`) and the decreasing variable( e.g. `[3, 3, 3, 2, 2, 1, 1]`)

**Comment by JoelJaeschke** [comment]
Hi @dcherian 
that was so obvious that I missed it..My bad! 

I am not sure if I understand correctly what you are saying, but if I implement the test as you describe, would this not already work? It will take the slow-path, but the behavior should be identical, right? My intention when writing the test was to explicitly test this case by supplying a grouper.

**Comment by dcherian** [comment]
TO check the fastpath we'd want to both get the right answer and make sure that `_group_indices` is filled with slices:
https://github.com/pydata/xarray/blob/4f3128bb7266d1a6fdeaf20718ea3f8286fe041e/xarray/core/groupby.py#L422-L430

**Comment by JoelJaeschke** [comment]
Hi @dcherian, I implemented the feedback as you described, hope this does the trick.

File: xarray/core/groupby.py [hunk 1]
```diff
@@ -400,7 +402,9 @@ def __init__(
             index = safe_cast_to_index(group)
             if not index.is_monotonic_increasing:
                 # TODO: sort instead of raising an error
-                raise ValueError("index must be monotonic for resampling")
```
Comment (dcherian) [inline xarray/core/groupby.py:403]: I undid this change. It's more involved to make it actually work.

File: xarray/tests/test_groupby.py [hunk 2]
```diff
@@ -1469,6 +1469,20 @@ def test_groupby_fillna(self):
...
 
+    def test_groupby_fastpath_for_monotonic(self):
+        # Fixes https://github.com/pydata/xarray/issues/6220
+        index = [1, 2, 3, 4, 7, 9, 10]
+        array = DataArray(np.arange(len(index)), [("idx", index)])
+        array_rev = array.copy().assign_coords({"idx": index[::-1]})
+        fwd = array.groupby("idx", squeeze=False)
+        rev = array_rev.groupby("idx", squeeze=False)
+
+        for gb in [fwd, rev]:
```
Comment (dcherian) [inline xarray/tests/test_groupby.py:1480]: nice test!

I just added a value check on the result below.

File: xarray/tests/test_groupby.py [hunk 3]
```diff
@@ -1490,9 +1504,13 @@ def test_resample(self):
         expected = DataArray.from_series(expected_)
         assert_identical(actual, expected)
 
-        with pytest.raises(ValueError, match=r"index must be monotonic"):
+        with pytest.raises(ValueError, match=r"Index must be monotonic"):
             array[[2, 0, 1]].resample(time="1D")
 
+        reverse = array.isel(time=slice(-1, None, -1))
+        with pytest.raises(ValueError):
```
Comment (dcherian) [inline xarray/tests/test_groupby.py:1511]: new test to make sure we error when trying to resample with a decreasing time index.

**Review by dcherian — APPROVED** [review]
Thanks @JoelJaeschke Can you add a line to `doc/whats-new.rst` please?

**Comment by JoelJaeschke** [comment]
Hi @dcherian, I added a note to whats-new. Thanks for your patience, still got a lot to learn :D

**Comment by Illviljan** [comment]
This one is failing on CI / ubuntu-latest py3.9 bare-minimum, is it taking a path without flox installed perhaps?

```
___________ TestDataArrayGroupBy.test_groupby_fastpath_for_monotonic ___________
[gw2] linux -- Python 3.9.16 /home/runner/micromamba-root/envs/xarray-tests/bin/python

self = <xarray.tests.test_groupby.TestDataArrayGroupBy object at 0x7fbd9f461c10>

    def test_groupby_fastpath_for_monotonic(self):
        # Fixes https://github.com/pydata/xarray/issues/6220
        index = [1, 2, 3, 4, 7, 9, 10]
        array = DataArray(np.arange(len(index)), [("idx", index)])
        array_rev = array.copy().assign_coords({"idx": index[::-1]})
        fwd = array.groupby("idx", squeeze=False)
        rev = array_rev.groupby("idx", squeeze=False)
    
        for gb in [fwd, rev]:
            assert all([isinstance(elem, slice) for elem in gb._group_indices])
    
        assert_identical(fwd.sum(), array)
>       assert_identical(rev.sum(), array_rev.sortby("idx"))
E       AssertionError: Left and right DataArray objects are not identical
E       
E       Differing values:
E       L
E           array([0, 1, 2, 3, 4, 5, 6])
E       R
E           array([6, 5, 4, 3, 2, 1, 0])
E       Differing coordinates:
E       L * idx      (idx) int64 10 9 7 4 3 2 1
E       R * idx      (idx) int64 1 2 3 4 7 9 10
```

File: xarray/tests/test_groupby.py [hunk 4]
```diff
@@ -1792,6 +1792,23 @@ def test_groupby_fillna(self) -> None:
         actual = a.groupby("b").fillna(DataArray([0, 2], dims="b"))
         assert_identical(expected, actual)
 
+    @pytest.mark.parametrize("use_flox", [True, False])
+    def test_groupby_fastpath_for_monotonic(self, use_flox):
```
Comment (headtr1ck) [inline xarray/tests/test_groupby.py:1796]:
```suggestion
    def test_groupby_fastpath_for_monotonic(self, use_flox: bool) -> None:
```

**Review by headtr1ck — APPROVED** [review]

**Review by max-sixty — APPROVED** [review]

**Comment by dcherian** [comment]
Thanks @JoelJaeschke sorry for the really long delay here!

**Pull request merged**